src/async_reader/mod.rs

```diff
@@ -473,6 +478,198 @@ impl<T: for<'a> ZarrReadAsync<'a> + Clone + Unpin + Send + 'static>
     }
 }
 
+//***********************************************************
+// implementation of an async stream that doesn't block when decompressing chunks. for now, that doesn't
+// to help at all, I need to revisit this at some point, won't be used for now. Also, for now filters are
+// not supported, i needed to add a "wrapper" for the mask stream, that packagaes the stream and the returned
+// chunk into a future, to avoid static lifetime issues.
+//************************************************************
+
+type InterleavedResults<T> = (
+    Result<StoreReadResults<T>, tokio::task::JoinError>,
+    Result<ZarrResult<RecordBatch>, tokio::task::JoinError>,
+);
+enum ZarrStreamStateNonBlocking<T: ZarrStream> {
+    Init,
+    Reading(BoxFuture<'static, StoreReadResults<T>>),
+    _Processing(ZarrRecordBatchReader<ZarrInMemoryChunkContainer>),
+    Interleaving(Option<ZarrRecordBatchReader<ZarrInMemoryChunkContainer>>),
+    ProcessingInterleaved(BoxFuture<'static, InterleavedResults<T>>),
+    Done,
+    Error,
+}
+
+pub struct ZarrRecordBatchStreamNonBlocking<'a, T: ZarrStream> {
+    meta: ZarrStoreMetadata,
+    filter: Option<ZarrChunkFilter>,
+    store_wrapper: Option<ZarrStoreWrapper<T>>,
+    state: ZarrStreamStateNonBlocking<T>,
+
+    // this is an optional record batch stream that will provide masks
+    // to apply to the data.
+    _mask_stream: Option<BoxStream<'a, ZarrResult<RecordBatch>>>,
+    _mask: Option<BooleanArray>,
+}
+
+impl<'a, T: ZarrStream> ZarrRecordBatchStreamNonBlocking<'a, T> {
+    fn new(meta: ZarrStoreMetadata, filter: Option<ZarrChunkFilter>, store: T) -> Self {
+        Self {
+            meta,
+            filter,
+            store_wrapper: Some(ZarrStoreWrapper::new(store)),
+            state: ZarrStreamStateNonBlocking::Init,
+            _mask_stream: None,
+            _mask: None,
+        }
+    }
+}
+
+impl<'a, T> Stream for ZarrRecordBatchStreamNonBlocking<'a, T>
+where
+    T: ZarrStream + Unpin + Send + 'static,
+{
+    type Item = ZarrResult<RecordBatch>;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
This function has a few `.expects` and `.unwraps`. I get some aren't possible, but in those cases perhaps we can follow something more idiomatic?